lib/mapping/MapEditUtils.cpp

```diff
@@ -0,0 +1,371 @@
+{
+	standardRule = (TerrainViewPattern::RULE_ANY == Name || TerrainViewPattern::RULE_DIRT == Name
+		|| TerrainViewPattern::RULE_NATIVE == Name || TerrainViewPattern::RULE_SAND == Name
+		|| TerrainViewPattern::RULE_TRANSITION == Name || TerrainViewPattern::RULE_NATIVE_STRONG == Name);
+	anyRule = (Name == TerrainViewPattern::RULE_ANY);
+	dirtRule = (Name == TerrainViewPattern::RULE_DIRT);
+	sandRule = (Name == TerrainViewPattern::RULE_SAND);
+	transitionRule = (Name == TerrainViewPattern::RULE_TRANSITION);
+	nativeStrongRule = (Name == TerrainViewPattern::RULE_NATIVE_STRONG);
+	nativeRule = (Name == TerrainViewPattern::RULE_NATIVE);
+}
+
+void TerrainViewPattern::WeightedRule::setNative()
+{
+	nativeRule = true;
+	standardRule = true;
+	//TODO: would look better as a bitfield
+	dirtRule = sandRule = transitionRule = nativeStrongRule = anyRule = false; //no idea what they mean, but look mutually exclusive
+}
+
+CTerrainViewPatternConfig::CTerrainViewPatternConfig()
+{
+	const JsonNode config(ResourceID("config/terrainViewPatterns.json"));
+	static const std::string patternTypes[] = { "terrainView", "terrainType" };
+	for (int i = 0; i < ARRAY_COUNT(patternTypes); ++i)
+	{
+		const auto& patternsVec = config[patternTypes[i]].Vector();
+		for (const auto& ptrnNode : patternsVec)
+		{
+			TerrainViewPattern pattern;
+
+			// Read pattern data
+			const JsonVector& data = ptrnNode["data"].Vector();
+			assert(data.size() == 9);
+			for (int j = 0; j < data.size(); ++j)
+			{
+				std::string cell = data[j].String();
+				boost::algorithm::erase_all(cell, " ");
+				std::vector<std::string> rules;
+				boost::split(rules, cell, boost::is_any_of(","));
+				for (std::string ruleStr : rules)
+				{
+					std::vector<std::string> ruleParts;
+					boost::split(ruleParts, ruleStr, boost::is_any_of("-"));
+					TerrainViewPattern::WeightedRule rule(ruleParts[0]);
+					assert(!rule.name.empty());
+					if (ruleParts.size() > 1)
+					{
+						rule.points = boost::lexical_cast<int>(ruleParts[1]);
+					}
+					pattern.data[j].push_back(rule);
+				}
+			}
+
+			// Read various properties
+			pattern.id = ptrnNode["id"].String();
+			assert(!pattern.id.empty());
+			pattern.minPoints = static_cast<int>(ptrnNode["minPoints"].Float());
+			pattern.maxPoints = static_cast<int>(ptrnNode["maxPoints"].Float());
+			if (pattern.maxPoints == 0) pattern.maxPoints = std::numeric_limits<int>::max();
+
+			// Read mapping
+			if (i == 0)
+			{
+				const auto& mappingStruct = ptrnNode["mapping"].Struct();
+				for (const auto& mappingPair : mappingStruct)
+				{
+					TerrainViewPattern terGroupPattern = pattern;
+					auto mappingStr = mappingPair.second.String();
+					boost::algorithm::erase_all(mappingStr, " ");
+					auto colonIndex = mappingStr.find_first_of(":");
+					const auto& flipMode = mappingStr.substr(0, colonIndex);
+					terGroupPattern.diffImages = TerrainViewPattern::FLIP_MODE_DIFF_IMAGES == &(flipMode[flipMode.length() - 1]);
```

Comment:
why is string compared to a pointer?

Reply:
@DjWarmonger , do you have an idea?

Reply:
Nah, it's string to string comparison and the result is stored in bool.

Reply:
![image](https://user-images.githubusercontent.com/2566990/190628895-359c4919-ceb1-4d2e-a17d-50dc80d27cfa.png)

But why is the pointer extracted from reference? No idea.

Reply:
then better to write like this:
```suggestion
					terGroupPattern.diffImages = TerrainViewPattern::FLIP_MODE_DIFF_IMAGES == flipMode.substr(flipMode.length() - 1, 1);
```

Reply:
Add braces around this expression